### PR TITLE
Prevent merging nodes' data paths

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today Elasticsearch does not prevent you from reconfiguring a node's
`path.data` to point to data paths that previously belonged to more than one
node. There's no good reason to be able to do this, and the consequences can be
quietly disastrous. Furthermore, #42489 might result in a user trying to split
up a previously-shared collection of data paths by hand and there's definitely
scope for mixing the paths up across nodes when doing this.

This change adds a check during startup to ensure that each data path belongs
to the same node.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 1]
```diff
@@ -494,10 +495,25 @@ private void maybeLogHeapDetails() {
     private static NodeMetaData loadOrCreateNodeMetaData(Settings settings, Logger logger,
                                                          NodePath... nodePaths) throws IOException {
         final Path[] paths = Arrays.stream(nodePaths).map(np -> np.path).toArray(Path[]::new);
+
+        final Set<String> nodeIds = new HashSet<>();
+        for (final Path path : paths) {
+            final NodeMetaData metaData = NodeMetaData.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY, path);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: Previously, a multi-path setup would succeed loading node state, as long as one of the paths were readable. With this change, I think we will fail if any of the paths are not readable/corrupt? See MetaDataStateFormat.loadGeneration(...)

**Review by henningandersen — COMMENTED** [review]
Looking good, my only concern is if the change in failure handling is important?

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: This is true, and I prefer the added strictness, but I've marked this as `team-discuss` so we can all make sure we're in agreement.

**Review by henningandersen — APPROVED** [review]
Discussed lenient vs strict reading on another channel and we decided that being stricter here was fine.

LGTM.

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/2

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**